nomad/structs/node.go

```diff
@@ -357,7 +357,7 @@ func (di *DriverInfo) HealthCheckEquals(other *DriverInfo) bool {
 
 // NodeMetaApplyRequest is used to update Node metadata on Client agents.
 type NodeMetaApplyRequest struct {
-	WriteRequest
+	QueryOptions // Client RPCs must use QueryOptions to set AllowStale=true
```

Comment:
Are there any upgrade path concern with this change? For example, a follower that has not been upgraded yet may try to forward a request to the a leader that _has_ been upgraded,  and so the request will have `WriteRequest` while the leader expects `QueryOptions`.

I haven't been able to break anything, so it seems fine,but I thought it would be worth double checking.

Reply:
> Are there any upgrade path concern with this change?

Great question. Concerns? No, I don't think so. It only effects in flight requests, not state, which is a nice start. The worst case scenario should be failed requests in mixed clusters, but in manual testing everything seemed to work fine!

The HTTP API isn't changed, so old CLIs will work with new HTTP agents and vice versa. This also means HTTP requests to the local Node are unaffected (eg via Task API), but they weren't impacted by the original bug anyway.

Our RPC framework must flatten QueryOptions and WriteRequest fields into the parent struct because when testing in a mixed server cluster (2 1.5.2-dev, 1 1.5.1), the requests always went through without error or infinite loop regardless of which server received the request or which server the client was connected to. So even in a mixed cluster where a 1.5.1 server improperly forwards a NodeMeta RPC to the leader, the leader, if 1.5.2, will set AllowStale=true which will be honored by servers of any version.

I had assumed a more pessimistic circumstance where QueryOptions.Region would appear empty and cause requests from old agents to appear empty, but that doesn't seem to be the case!